src/backend/distributed/master/master_metadata_utility.c

```diff
@@ -62,6 +64,274 @@ static ShardPlacement * TupleToShardPlacement(TupleDesc tupleDesc,
+	appendStringInfo(selectQuery, "0;");
+	return selectQuery;
+}
+
+
+/*
+ * SupportFunctionForCitusSize is a helper function for all citus size functions. It
+ * first checks whether the table is distributed and size query can be run on it.
+ * Connection to each node has to be established to get the size of the table.
+ */
+uint64
+SupportFunctionForCitusSize(Oid relationId, char *sizeQuery)
+{
+	MultiConnection *connection = NULL;
+	List *workerNodeList = WorkerNodeList();
+	ListCell *workerNodeCell = NULL;
+	uint64 totalRelationSize = 0;
+
+	/* check whether the size of table can be obtained with citus size functions */
+	ErrorIfNotSuitableToGetSize(relationId);
+
+	/* accessing single placement is enough for reference tables */
+	if (PartitionMethod(relationId) == DISTRIBUTE_BY_NONE)
+	{
+		PGresult *queryResult = NULL;
+		StringInfo tableSizeQuery = NULL;
+		ListCell *referenceWorkerNodeCell = list_head(workerNodeList);
+		WorkerNode *referenceWorkerNode = (WorkerNode *) lfirst(referenceWorkerNodeCell);
+
+		char *referenceWorkerNodeName = referenceWorkerNode->workerName;
+		uint32 referenceWorkerNodePort = referenceWorkerNode->workerPort;
+		char *tableSizeString;
+		int executeCommand = 0;
+		uint32 connectionFlag = 0;
+		uint64 currentRelationSize = 0;
+
+		List *intervalsOnNode = GetShardIntervalsOnNode(referenceWorkerNode, relationId);
+
+		tableSizeQuery = GenerateSizeQueryOnMultiplePlacements(relationId,
+															   intervalsOnNode,
+															   sizeQuery);
+
+		connection = GetNodeConnection(connectionFlag,
+									   referenceWorkerNodeName,
+									   referenceWorkerNodePort);
+
+		executeCommand = ExecuteOptionalRemoteCommand(connection, tableSizeQuery->data,
+													  &queryResult);
+
+		/* If there is a connection error close the connection and quit */
+		if (executeCommand != 0)
+		{
+			ReportConnectionError(connection, ERROR);
+		}
+
+		tableSizeString = PQgetvalue(queryResult, 0, 0);
+		currentRelationSize = atol(tableSizeString);
+
+		PQclear(queryResult);
+
+		ForgetResults(connection);
+
+		totalRelationSize = list_length(workerNodeList) * currentRelationSize;
+		return totalRelationSize;
+	}
+
+	foreach(workerNodeCell, workerNodeList)
+	{
+		PGresult *queryResult = NULL;
+		StringInfo tableSizeQuery = NULL;
+		WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
+		char *workerNodeName = workerNode->workerName;
+		uint32 workerNodePort = workerNode->workerPort;
+		char *tableSizeString;
+		int executeCommand = 0;
+		uint32 connectionFlag = 0;
+		uint64 currentRelationSize = 0;
+
+		List *intervalsOnNode = GetShardIntervalsOnNode(workerNode, relationId);
+
+		tableSizeQuery = GenerateSizeQueryOnMultiplePlacements(relationId,
+															   intervalsOnNode,
+															   sizeQuery);
+
+		connection = GetNodeConnection(connectionFlag, workerNodeName, workerNodePort);
+
+		executeCommand = ExecuteOptionalRemoteCommand(connection, tableSizeQuery->data,
```

Comment:
The scenario is a little weird, but transaction blocks that first add a column and then perform `citus_table_size` will typically deadlock, since `pg_table_size` blocks when a table is being altered. We might want to prevent this scenario.

For example, we could add an `if (XactModificationLevel == XACT_MODIFICATION_MULTI_SHARD)` check at the start of this function.